File: python/akf/builder.py
"""AKF v1.0 — Fluent builder API."""

import uuid
from datetime import datetime, timezone
from typing import List, Optional

from .models import AKF, Claim, ProvHop
from .provenance import compute_integrity_hash


class AKFBuilder:
    """Fluent builder for constructing AKF units."""

    def __init__(self) -> None:
        self._claims: List[Claim] = []
        self._by: Optional[str] = None
        self._agent_id: Optional[str] = None
        self._label: Optional[str] = None
        self._inherit: Optional[bool] = None
        self._ext: Optional[bool] = None
        self._ttl: Optional[int] = None
        self._meta: Optional[dict] = None
        self._model: Optional[str] = None
        self._tools: Optional[List[str]] = None
        self._session: Optional[str] = None

    def claim(self, content: str, trust: float, **kwargs) -> "AKFBuilder":
        """Add a claim."""
        self._claims.append(Claim(content=content, confidence=trust, **kwargs))
        return self

    def by(self, author: str) -> "AKFBuilder":
        """Set author."""
        self._by = author
        return self

    def agent(self, agent_id: str) -> "AKFBuilder":
        """Set AI agent."""
        self._agent_id = agent_id
        return self

    def label(self, classification: str) -> "AKFBuilder":
        """Set security classification."""
        self._label = classification
        return self

    def inherit(self, value: bool = True) -> "AKFBuilder":
        """Set inheritance flag."""
        self._inherit = value
        return self

    def ext(self, value: bool = True) -> "AKFBuilder":
        """Set external sharing flag."""
        self._ext = value
        return self

    def ttl(self, days: int) -> "AKFBuilder":
        """Set retention period in days."""
        self._ttl = days
        return self

    def tag(self, *tags: str) -> "AKFBuilder":
        """Add tags to the last claim."""
        if not self._claims:
            raise ValueError("No claims to tag — add a claim first")
        last = self._claims[-1]
        existing = list(last.tags) if last.tags else []
        existing.extend(tags)
        self._claims[-1] = last.model_copy(update={"tags": existing})
        return self

    def meta(self, **kwargs) -> "AKFBuilder":
        """Set free-form metadata."""
        self._meta = kwargs
        return self

    def kind(self, kind: str) -> "AKFBuilder":
        """Set the kind on the last claim."""
        if not self._claims:
            raise ValueError("No claims — add a claim first")
        last = self._claims[-1]
        self._claims[-1] = last.model_copy(update={"kind": kind})
        return self

    def evidence(self, *items) -> "AKFBuilder":
        """Add evidence to the last claim. Accepts Evidence objects, dicts, or strings."""
        if not self._claims:
            raise ValueError("No claims — add a claim first")
        from .stamp import parse_evidence_string
        from .models import Evidence

        parsed = []
        for item in items:
            if isinstance(item, Evidence):
                parsed.append(item)
            elif isinstance(item, dict):
                parsed.append(Evidence(**item))
            elif isinstance(item, str):
                parsed.append(parse_evidence_string(item))
            else:
                raise TypeError(f"Unsupported evidence type: {type(item)}")

        last = self._claims[-1]
        existing = list(last.evidence) if last.evidence else []
        existing.extend(parsed)
        self._claims[-1] = last.model_copy(update={"evidence": existing})
        return self

    def model(self, model_id: str) -> "AKFBuilder":
        """Set the model identifier."""
        self._model = model_id
        return self

    def tools(self, *tool_names: str) -> "AKFBuilder":
        """Set the tools used."""
        self._tools = list(tool_names)
        return self

    def session(self, session_id: str) -> "AKFBuilder":
        """Set the session identifier."""
        self._session = session_id
        return self

    def build(self) -> AKF:
        """Build the AKF unit."""
        if not self._claims:
            raise ValueError("At least one claim is required")

        now = datetime.now(timezone.utc).isoformat()
        unit_id = "akf-{}".format(uuid.uuid4().hex[:12])

        # Auto-create provenance hop 0
        prov: Optional[List[ProvHop]] = None
        if self._by or self._agent_id:
            actor = self._by or self._agent_id or "unknown"
            prov = [
                ProvHop(
                    hop=0,
                    actor=actor,
                    action="created",
                    timestamp=now,
                    claims_added=[c.id for c in self._claims if c.id],
                )
            ]

        unit = AKF(
            version="1.0",
            id=unit_id,
            claims=self._claims,
            author=self._by,
            agent=self._agent_id,
            model=self._model,
            tools=self._tools,
            session=self._session,
            created=now,
            classification=self._label if self._label is not None else "internal",
            inherit_classification=self._inherit if self._inherit is not None else True,
            allow_external=self._ext if self._ext is not None else False,
            ttl=self._ttl,
            prov=prov,
            meta=self._meta,
        )

        # Auto-compute integrity hash
        integrity = compute_integrity_hash(unit)
        unit = unit.model_copy(update={"integrity_hash": integrity})

        return unit

File: typescript/src/builder.ts
/**
 * AKF v1.1 — Fluent builder API.
 */

import { randomUUID } from "node:crypto";
import type { AKFUnit, Claim, Evidence, ProvHop } from "./models.js";
import { computeIntegrityHash } from "./provenance.js";

/** Fluent builder for constructing AKF units. */
export class AKFBuilder {
  private _claims: Claim[] = [];
  private _by: string | undefined;
  private _agentId: string | undefined;
  private _modelId: string | undefined;
  private _label: string | undefined;
  private _inherit: boolean | undefined;
  private _ext: boolean | undefined;
  private _ttl: number | undefined;
  private _tools: string[] | undefined;
  private _session: string | undefined;
  private _meta: Record<string, unknown> | undefined;

  /** Add a claim. */
  claim(
    content: string,
    trust: number,
    opts?: Partial<Omit<Claim, "c" | "t">>
  ): this {
    const claimId: string = (opts?.id as string) || randomUUID().replace(/-/g, "").slice(0, 8);
    this._claims.push({ c: content, t: trust, ...opts, id: claimId });
    return this;
  }

  /** Set author. */
  by(author: string): this {
    this._by = author;
    return this;
  }

  /** Set AI agent identifier. */
  agent(id: string): this {
    this._agentId = id;
    return this;
  }

  /** Set security classification. */
  label(classification: string): this {
    this._label = classification;
    return this;
  }

  /** Set inheritance flag. */
  inherit(value: boolean = true): this {
    this._inherit = value;
    return this;
  }

  /** Set external sharing flag. */
  ext(value: boolean = true): this {
    this._ext = value;
    return this;
  }

  /** Set retention period in days. */
  ttl(days: number): this {
    this._ttl = days;
    return this;
  }

  /** Set model identifier. */
  model(id: string): this {
    this._modelId = id;
    return this;
  }

  /** Set kind on the last claim. */
  kind(kind: string): this {
    if (this._claims.length === 0) {
      throw new Error("No claims to set kind on — add a claim first");
    }
    const last = this._claims[this._claims.length - 1];
    this._claims[this._claims.length - 1] = { ...last, kind };
    return this;
  }

  /** Add evidence to the last claim. */
  evidence(...items: Evidence[]): this {
    if (this._claims.length === 0) {
      throw new Error("No claims to add evidence to — add a claim first");
    }
    const last = this._claims[this._claims.length - 1];
    const existing = last.evidence ? [...last.evidence] : [];
    existing.push(...items);
    this._claims[this._claims.length - 1] = { ...last, evidence: existing };
    return this;
  }

  /** Set tools used. */
  tools(...names: string[]): this {
    this._tools = names;
    return this;
  }

  /** Set session identifier. */
  session(id: string): this {
    this._session = id;
    return this;
  }

  /** Add tags to the last claim. */
  tag(...tags: string[]): this {
    if (this._claims.length === 0) {
      throw new Error("No claims to tag — add a claim first");
    }
    const last = this._claims[this._claims.length - 1];
    const existing = last.tags ? [...last.tags] : [];
    existing.push(...tags);
    this._claims[this._claims.length - 1] = { ...last, tags: existing };
    return this;
  }

  /** Set free-form metadata. */
  meta(data: Record<string, unknown>): this {
    this._meta = data;
    return this;
  }

  /** Build the AKF unit. */
  build(): AKFUnit {
    if (this._claims.length === 0) {
      throw new Error("At least one claim is required");
    }

    const now = new Date().toISOString();
    const id = `akf-${randomUUID().replace(/-/g, "").slice(0, 12)}`;

    // Auto-create provenance hop 0
    let prov: ProvHop[] | undefined;
    if (this._by || this._agentId) {
      const actor = this._by || this._agentId || "unknown";
      prov = [
        {
          hop: 0,
          by: actor,
          do: "created",
          at: now,
          adds: this._claims
            .map((c) => c.id)
            .filter((id): id is string => id !== undefined),
        },
      ];
    }

    const unit: AKFUnit = {
      v: "1.0",
      id,
      claims: this._claims,
      by: this._by,
      agent: this._agentId,
      model: this._modelId,
      tools: this._tools,
      session: this._session,
      at: now,
      label: this._label,
      inherit: this._inherit,
      ext: this._ext,
      ttl: this._ttl,
      prov,
      meta: this._meta,
    };

    // Auto-compute integrity hash
    const integrity = computeIntegrityHash(unit);
    unit.hash = integrity;

    return unit;
  }
}
